examples/psyad/lfric/adjoint/adj_poly1d_vert_adv_kernel_mod.F90

```diff
@@ -0,0 +1,125 @@
+    integer(kind=i_def), dimension(ndf_c), intent(in) :: map_c
+    integer(kind=i_def), intent(in) :: ndata
+    integer(kind=i_def), intent(in) :: global_order
+    real(kind=r_def), dimension(undf_wt), intent(in) :: advective
+    real(kind=r_def), dimension(undf_w2), intent(inout) :: wind
+    real(kind=r_def), dimension(undf_wt), intent(inout) :: tracer
+    real(kind=r_def), dimension(undf_w2), intent(in) :: ls_wind
+    real(kind=r_def), dimension(undf_wt), intent(in) :: ls_tracer
+    real(kind=r_def), dimension(undf_c), intent(in) :: coeff
+    logical(kind=l_def), intent(in) :: logspace
+    integer(kind=i_def) :: k
+    integer(kind=i_def) :: kmin
+    integer(kind=i_def) :: kmax
+    integer(kind=i_def) :: ij
+    integer(kind=i_def) :: ik
+    integer(kind=i_def) :: p
+    integer(kind=i_def) :: vertical_order
+    integer(kind=i_def) :: use_upwind
+    integer(kind=i_def) :: upwind_offset
+    integer(kind=i_def) :: upwind
+    INTEGER(KIND = i_def), DIMENSION(global_order + 1) :: stencil
+    real(kind=r_def) :: dpdz
+    real(kind=r_def) :: ls_dpdz
+    real(kind=r_def), dimension(0:nlayers) :: ls_log_tracer
+    real(kind=r_def) :: tmp1
+    real(kind=r_def) :: tmp2
+    real(kind=r_def) :: tmp3
+    integer(kind=i_def) :: i
+
+    dpdz = 0.0_r_def
+    ij = map_wt(1)
+    if (logspace) then
+      do k = 0, nlayers, 1
+        ls_log_tracer(k) = LOG(ABS(ls_tracer(ij + k)))
+      enddo
+    end if
+    vertical_order = MIN(global_order, nlayers - 1)
+    use_upwind = MOD(vertical_order, 2)
+    do k = nlayers - 1, 1, -1
+      do p = 0, vertical_order, 1
+        tmp1 = REAL(vertical_order, r_def)
+        tmp2 = 0.5 * tmp1
+        tmp3 = FLOOR(tmp2)
+        stencil(p + 1) = k + p - tmp3
+      enddo
+      upwind = INT(0.5 * SIGN(1.0, ls_wind(k + map_w2(5))) + 0.5, i_def)
+      upwind_offset = upwind * use_upwind
+      stencil = stencil - upwind_offset
+      kmin = stencil(1)
+      do i = 2, vertical_order + 1, 1
+        if (stencil(i) < kmin) then
+          kmin = stencil(i)
+        end if
+      enddo
+      if (kmin < 0) then
+        stencil = -kmin + stencil
+      end if
+      kmax = stencil(1)
+      do i = 2, vertical_order + 1, 1
+        if (stencil(i) > kmax) then
+          kmax = stencil(i)
+        end if
+      enddo
+      kmax = kmax - nlayers
+      if (kmax > 0) then
+        stencil = -kmax + stencil
+      end if
+      ls_dpdz = 0.0_r_def
+      dpdz = dpdz + advective(map_wt(1) + k) * ls_wind(k + map_w2(5))
+      wind(k + map_w2(5)) = wind(k + map_w2(5)) + ls_dpdz * advective(map_wt(1) + k)
+      if (logspace) then
+        ls_dpdz = ls_dpdz * ls_tracer(ij + k)
+        tracer(ij + k) = tracer(ij + k) + ls_dpdz * dpdz
+        dpdz = dpdz * ls_tracer(ij + k)
+        do p = vertical_order + 1, 1, -1
+          ik = global_order * upwind_offset + k * ndata + p + upwind_offset + map_c(1) - 1
+          ls_dpdz = ls_dpdz + coeff(ik) * ls_log_tracer(stencil(p))
```

Comment:
This code isn't correct as we need to calculate the passive `ls_dpdz` first before then looping to calculate `tracer` which is active. I think the simplest solution is simply to copy all the kernels out of my updated `jan23_adj_testing` branch.

Reply:
Copied files as suggested. However, in a couple of cases there was any_space_9 metadata instead of wchi, so I changed that. There was also one error in lbl_adj_poly2d_w3_reconstruction_kernel
_mod.F90 where the declaration of polynomial_tracer was nalyers when it should have been 0:nlayers-1.